lib/RestructureCFG/InlineDivergentScopesPass.cpp

```diff
@@ -334,24 +376,123 @@ static void performIDS(DivergenceDescriptor &DivergenceDescriptor,
   simplifyTerminator(Tail, PlaceHolderTarget);
 
   // We add a `scope_closer` edge between the divergent exit node and
-  // the `Tail` node
-  ScopeGraphBuilder SGBuilder(F);
-  SGBuilder.addScopeCloser(Exit, Tail);
+  // the `Tail` node for all the exits
+  for (auto Divergence : MultipleDivergences) {
+    SGBuilder.addScopeCloser(Divergence.Exit, Tail);
+  }
+}
+
+/// Helper function which returns if a `DivergenceDescriptor` is compatible with
+/// all the ones already present in a `Collection`. Being compatible means that
+/// they insist on the same conditional, have non-overlapping `Exit` blocks, and
+/// do not have any overlapping `DivergentSuccessors`
+static bool
+canAddDescriptor(const SmallVector<DivergenceDescriptor> &Collection,
+                 const DivergenceDescriptor &NewDescriptor) {
+  // Skip checking if collection is empty
+  if (Collection.empty())
+    return true;
+
+  // Check that Conditional matches
+  if (Collection[0].Conditional != NewDescriptor.Conditional)
+    return false;
```

Comment:
Can this even happen?

Reply:
Yes, we have no guarantee that the divergence we are trying to add insists on the same conditional. More on this in the following comment explaining the rationale behind `canAddDescriptor`.